tests/e2e/ctl_v3_member_test.go

```diff
@@ -304,3 +312,229 @@ func TestRemoveNonExistingMember(t *testing.T) {
 	// Ensure that membership is properly bootstrapped.
 	assert.NoError(t, epc.Restart(ctx))
 }
+
+// TestCtlV3PromotingLearner tests whether etcd can automatically fix
+// the issue caused by https://github.com/etcd-io/etcd/issues/19557
+// when upgrading from 3.5 to 3.6.
+func TestCtlV3PromotingLearner(t *testing.T) {
+	testCases := []struct {
+		name                  string
+		snapshotCount         uint64
+		writeToV3StoreSuccess bool
+	}{
+		{
+			name:          "create snapshot after learner promotion which is not saved to v3store",
+			snapshotCount: 10,
+		},
+		{
+			name:          "no snapshot and learner promotion is not saved to v3store",
+			snapshotCount: 0,
+		},
+		{
+			name:                  "no snapshot and learner promotion is saved to v3store",
+			snapshotCount:         0,
+			writeToV3StoreSuccess: true,
+		},
+	}
+
+	if !fileutil.Exist(e2e.BinPath.EtcdLastRelease) {
+		t.Skipf("%q does not exist", e2e.BinPath.EtcdLastRelease)
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			e2e.BeforeTest(t)
+
+			ctx := context.Background()
+
+			epc, promotedMembers := mustCreateNewClusterByPromotingMembers(t, e2e.LastVersion, 2, e2e.WithSnapshotCount(tc.snapshotCount))
+			defer func() {
+				require.NoError(t, epc.Close())
+			}()
+			require.Len(t, promotedMembers, 1)
+			t.Logf("Promoted member: %+v", promotedMembers[0])
+
+			if tc.snapshotCount != 0 {
+				t.Logf("Write %d keys to trigger a snapshot", tc.snapshotCount)
+				for i := 0; i < int(tc.snapshotCount); i++ {
+					err := epc.Etcdctl().Put(ctx, fmt.Sprintf("key_%d", i), fmt.Sprintf("value_%d", i), config.PutOptions{})
+					require.NoError(t, err)
+				}
+			}
+
+			if tc.writeToV3StoreSuccess {
+				t.Log("Skip manually changing the already promoted member to a learner in v3store")
+			} else {
+				t.Logf("Stopping all members")
+				require.NoError(t, epc.Stop())
+
+				t.Log("Manually changing the already promoted member to a learner again in the v3store of all members to simulate the issue https://github.com/etcd-io/etcd/issues/19557")
+				promotedMembers[0].IsLearner = true
+				for i := 0; i < len(epc.Procs); i++ {
+					mustSaveMemberIntoBbolt(t, epc.Procs[i].Config().DataDirPath, promotedMembers[0])
+				}
+			}
+
+			t.Log("Upgrading the cluster")
```

Comment:
we can move this assignment to the beginning of the test 

```go
func TestCtlV3PromotingLearner(t *testing.T) {
  -- here
```

Reply:
not quite get the point. I intentionally print the log right before we upgrade the cluster.

Changed the test name to `TestClusterUpgradeWithPromotingLearner`

Reply:
sorry I comment on wrong line. I mean these four lines

```go
	                currentVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.Etcd)
			require.NoError(t, err)
			lastVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.EtcdLastRelease)
			require.NoError(t, err)
```

Reply:
I intentionally added them right before where they being used.

But overall seems not a big deal.